pkg/service/auth_context.go

```diff
@@ -23,44 +27,125 @@ type AuthContext struct {
 	Authorization map[*config.AuthorizationConfig]interface{}
 }
 
-// Evaluate evaluates all steps of the auth pipeline (identity → metadata → policy enforcement)
-func (self *AuthContext) Evaluate() error {
-	self.Identity = make(map[*config.IdentityConfig]interface{})
-	self.Metadata = make(map[*config.MetadataConfig]interface{})
-	self.Authorization = make(map[*config.AuthorizationConfig]interface{})
-
-	// identity (authentication)
-	identityConfigs := self.API.IdentityConfigs
-	for i := range identityConfigs {
-		c := identityConfigs[i]
-		if ret, err := c.Call(self); err != nil {
+// AuthObjectConfig provides an interface for APIConfig objects that implements a Call method
+type AuthObjectConfig interface {
+	Call(ctx common.AuthContext) (interface{}, error)
+}
+
+type configCallback = func(config AuthObjectConfig, obj interface{})
+
+// NewAuthContext creates an AuthContext instance
+func NewAuthContext(parentCtx context.Context, req *auth.CheckRequest, apiConfig config.APIConfig) AuthContext {
+
+	return AuthContext{
+		ParentContext: &parentCtx,
+		Request:       req,
+		API:           &apiConfig,
+		Identity:      make(map[*config.IdentityConfig]interface{}),
+		Metadata:      make(map[*config.MetadataConfig]interface{}),
+		Authorization: make(map[*config.AuthorizationConfig]interface{}),
+	}
+
+}
+
+func (authContext *AuthContext) getAuthObject(ctx context.Context, objConfig AuthObjectConfig, cb configCallback) error {
+	log := ctrl.Log.WithName("Authorino").WithName("AuthContext")
+	select {
+	case <-ctx.Done():
+		log.Info("Context cancelled objConfig %v terminting", objConfig)
+		return nil
+	default:
+		if authObj, err := objConfig.Call(authContext); err != nil {
+			log.Error(err, "Invalid auth object config")
 			return err
 		} else {
-			self.Identity[&c] = ret
-			break
+			cb(objConfig, authObj)
+			return nil
 		}
 	}
+}
+
+func (authContext *AuthContext) getAuthObjects(configs []AuthObjectConfig, cb configCallback) error {
+	log := ctrl.Log.WithName("Authorino").WithName("AuthContext")
+	errGroup, ctx := errgroup.WithContext(context.Background())
+
+	for _, config := range configs {
+		objConfig := config
+		errGroup.Go(func() error {
+			return authContext.getAuthObject(ctx, objConfig, cb)
+		})
+	}
+	if err := errGroup.Wait(); err != nil {
+		return err
+	} else {
+		log.Info("Successfully fetched all auth objects.", configs)
+		return nil
+	}
+}
+
+// GetIDObject gets an Identity auth object given an Identity config.
+func (authContext *AuthContext) GetIDObject() error {
+	configs := make([]AuthObjectConfig, len(authContext.API.IdentityConfigs))
+	// Convert []SpecificType to []interfaceType
+	for i, conf := range authContext.API.IdentityConfigs {
+		cpConf := conf
+		configs[i] = &cpConf
+	}
```

Comment:
I wish we had another way to avoid this conversion (here and in the other analogous steps below), but I guess there's nothing we can do about it... part of the hurdles of using Golang perhaps. @jmprusi @eguzki, any ideas?

Reply:
IMO. this is a bad smell coming from the data structure design. It should be reviewed and designed better with interfaces. I ti would allow easier unittesting as well.

Reply:
I'll try to see how could it be done in my following refactor PRs

Reply:
I think I kind of have a good fix for this, will commit it in this PR